### ISSUE TITLE
Convert guide to using WSL 2 + linux `brew` for all Windows users. 

### ISSUE THREAD

**Issue by brookr** [codefellows/setup-guide#19]
Use this thread to share your best practices, steps to configure, and hiccups along the way!

**Comment by JB-Tellez** [comment]
I haven't found any issues with installing WSL itself when following Microsoft official docs here ttps://docs.microsoft.com/en-us/windows/wsl/

**Comment by brookr** [comment]
[EDIT THIS NOTE AS NEEDED]

Things that we need to verify that work with WSL 2: 

201, 301 -
- brew
- nvm
- node 
- tree
- VSCode 

401:JS - 
- mongo 
- Expo

**Comment by johncokos** [comment]
Here's my full "How To" on getting everything I think we need going:

# WSL Setup

## Basic Updates
- sudo apt update
- sudo apt-get install build-essential

## Install linux brew
- /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)"
- test -d ~/.linuxbrew && eval $(~/.linuxbrew/bin/brew shellenv)
- test -d /home/linuxbrew/.linuxbrew && eval $(/home/linuxbrew/.linuxbrew/bin/brew shellenv)
- test -r ~/.bashrc && echo "eval \$($(brew --prefix)/bin/brew shellenv)" >>~/.bashrc

### Reset Env:
- source ~/.bashrc

## Install gcc
- brew install gcc

## Install nvm
- brew install nvm
- Add these lines to ~/.bashrc (`sudo nano ~/.bashrc`)
   ```
   export NVM_DIR="$HOME/.nvm"
  [ -s "/home/linuxbrew/.linuxbrew/opt/nvm/nvm.sh" ] && . "/home/linuxbrew/.linuxbrew/opt/nvm/nvm.sh"  # This loads nvm
  [ -s "/home/linuxbrew/.linuxbrew/opt/nvm/etc/bash_completion.d/nvm" ] && . "/home/linuxbrew/.linuxbrew/opt/nvm/etc/bash_completion.d/nvm"  # This loads nvm bash_completion
   ```

### Reset Env:
- source ~/.bashrc

## Node and Utilities
- nvm install 12
- npm install -g live-server json-server nodemon

## Python
- brew install python

## PHP
- brew install php

## Apache
- brew install httpd

## Install Tree
- brew install tree

## Install AWS and Azure CLI
- brew install awscli
- brew install azure-cli

## Postgres
- brew install postgresql
- pg_ctl -D /home/linuxbrew/.linuxbrew/var/postgres start
- createdb `whoami`

## Mongo
- wget -qO - https://www.mongodb.org/static/pgp/server-4.2.asc | sudo apt-key add -
- sudo apt-get install gnupg
- echo "deb [ arch=amd64,arm64 ] https://repo.mongodb.org/apt/ubuntu bionic/mongodb-org/4.2 multiverse" | sudo tee /etc/apt/sources.list.d/mongodb-org-4.2.list
- sudo apt-get update
- sudo apt-get install -y mongodb-org
- sudo nano /etc/init.d/mongod and put the contents from this:
    - https://raw.githubusercontent.com/mongodb/mongo/master/debian/init.d
- sudo chmod 755 /etc/init.d/mongod
- sudo service mongod start

## VSCODE

- Install it
- Instsall the `remote-wsl` extension

## Verify most of the goodness ....

Close all the way down (restart the computer if you want ...) then to back to Ubuntu and run these:

```
node --version
npm --version
python --version
php --version
live-server --version
json-server --version
tree --version
aws --version
azure --version
psql --version
mongo --version
mongod --version
```

### Issues

#### Database Services

It does not look like WSL2 can auto-start anything, so you'll always have to restart mongo and postgres. Once you do, the services survive sessions, just not a computer restart:

`sudo service mongod start`

`pg_ctl -D /home/linuxbrew/.linuxbrew/var/postgres start`


#### VSCode

I have the extensions installed, but I can't get `code .` to work in WSL2 ... no way that I can see to add that to your path and make that work. If someone knows how, I can update the above.

**Comment by johncokos** [comment]
One nice thing is postgres connection strings in WSL2 don't require the username and password. Just `postgres://localhost:5432/databasename` works fine as is.

**Comment by johncokos** [comment]
More Good Reportings ....

Running a React App on :3000

Tried to start it again and got this:

```
? Something is already running on port 3000. Probably:
  api-client (pid 2818)
  in /home/johncokos/codefellows/guides/code-401-javascript-guide/curriculum/apps-and-libraries/validator-app

Would you like to run the app on another port instead? (Y/n) npm ERR! code ELIFECYCLE
```

Create React App is good at that though but how about express?

```
[nodemon] starting `node app.js`
events.js:292
      throw er; // Unhandled 'error' event
      ^

Error: listen EADDRINUSE: address already in use :::3000
    at Server.setupListenHandle [as _listen2] (net.js:1313:16)
    at listenInCluster (net.js:1361:12)
    at Server.listen (net.js:1447:7)
    at Function.listen (/home/johncokos/codefellows/api/node_modules/express/lib/application.js:618:24)
    at Object.<anonymous> (/home/johncokos/codefellows/api/app.js:7:5)
    at Module._compile (internal/modules/cjs/loader.js:1138:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1158:10)
    at Module.load (internal/modules/cjs/loader.js:986:32)
    at Function.Module._load (internal/modules/cjs/loader.js:879:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:71:12)
Emitted 'error' event on Server instance at:
    at emitErrorNT (net.js:1340:8)
    at processTicksAndRejections (internal/process/task_queues.js:84:21) {
  code: 'EADDRINUSE',
  errno: 'EADDRINUSE',
  syscall: 'listen',
  address: '::',
  port: 3000
}
[nodemon] app crashed - waiting for file changes before starting...
```

Could we actually be in a world where students' won't have 50 ghost processes attaching to the same port?

**Comment by johncokos** [comment]
Lot of words up there ... but need help with VSCode and WSL2. I can't connect them.

**Comment by brookr** [comment]
@johncokos How did you install Expo?

**Comment by zeyad-kay** [comment]
> ## Mongo
> * `wget -qO - https://www.mongodb.org/static/pgp/server-4.2.asc | sudo apt-key add -`
> * `sudo apt-get install gnupg`
> * `echo "deb [ arch=amd64,arm64 ] https://repo.mongodb.org/apt/ubuntu bionic/mongodb-org/4.2 multiverse" | sudo tee /etc/apt/sources.list.d/mongodb-org-4.2.list`
> * `sudo apt-get update`
> * `sudo apt-get install -y mongodb-org`
> * `sudo nano /etc/init.d/mongod` and put the contents from this:z
>   
>   * https://raw.githubusercontent.com/mongodb/mongo/master/debian/init.d
> * `sudo chmod 755 /etc/init.d/mongod`
> * `sudo service mongod start`

I have tried this a couple of times but when I start mongod, it fails
`Starting database mongod                                              [fail]`

**Comment by dahlbyk** [comment]
Not sure if this should be a separate issue, but old 201 prework had students install Git on the Windows side for VS Code to use. Is that not needed with WSL Remote extension?

401 .NET prework will need to be updated to make sure Git is installed, unless they've updated the Visual Studio installer to actually install Git. (It didn't used to because GPL.)
